File: frontend/src/api/validation.ts
import { apiClient } from "./client";

export interface RegoValidationResult {
    valid: boolean;
    errors?: string[];
    warnings?: string[];
}

export interface RegoEvaluationResult {
    result: unknown;
    errors?: string[];
    warnings?: string[];
    command?: string;
}

export const validateRego = async (rego: string): Promise<RegoValidationResult> => {
    const response = await apiClient.post<RegoValidationResult>("/validation/rego", { rego });
    return response.data;
};

export const evaluateRego = async (
    rego: string,
    input: unknown
): Promise<RegoEvaluationResult> => {
    const response = await apiClient.post<RegoEvaluationResult>("/validation/rego/eval", { rego, input });
    return response.data;
};

File: backend/app/routers/validation.py
from __future__ import annotations

from fastapi import APIRouter, Depends
from ..schemas import (
    RegoEvaluationRequest,
    RegoEvaluationResponse,
    RegoValidationRequest,
    RegoValidationResponse,
)
from ..services.rego_validator import RegoValidator

router = APIRouter(prefix="/validation", tags=["validation"])


def get_validator() -> RegoValidator:
    return RegoValidator()


@router.post("/rego", response_model=RegoValidationResponse)
def validate_rego(
    payload: RegoValidationRequest,
    validator: RegoValidator = Depends(get_validator),
) -> RegoValidationResponse:
    valid, errors, warnings = validator.validate(payload.rego)
    return RegoValidationResponse(
        valid=valid, errors=errors or None, warnings=warnings or None
    )


@router.post("/rego/eval", response_model=RegoEvaluationResponse)
def evaluate_rego(
    payload: RegoEvaluationRequest,
    validator: RegoValidator = Depends(get_validator),
) -> RegoEvaluationResponse:
    result, errors, warnings, command = validator.evaluate(
        payload.rego, payload.input
    )
    return RegoEvaluationResponse(
        result=result,
        errors=errors or None,
        warnings=warnings or None,
        command=command or None,
    )
